apps/api-extractor/src/collector/MessageRouter.ts

```diff
@@ -222,6 +242,57 @@ export class MessageRouter {
     }
   }
 
+  /**
+   * Recursively collects the primitive members (numbers, strings, arrays, etc) into an object that
+   * is JSON serializable.  This is used by the "--diagnostics" feature to dump the state of configuration objects.
+   *
+   * @returns a JSON serializable object (possibly including `null` values)
+   *          or `undefined` if the input cannot be represented as JSON
+   */
+  // tslint:disable-next-line:no-any
+  public static buildJsonDumpObject(input: any): any | undefined {
+    if (input === null || input === undefined) {
+      // tslint:disable-next-line:no-null-keyword
+      return null; // JSON uses null instead of undefined
+    }
+
+    switch (typeof input) {
+      case 'boolean':
+      case 'number':
+      case 'string':
+        return input;
+      case 'object':
+        if (Array.isArray(input)) {
+          // tslint:disable-next-line:no-any
+          const outputArray: any[] = [];
+          for (const element of input) {
+            // tslint:disable-next-line:no-any
+            const serializedElement: any = MessageRouter.buildJsonDumpObject(element);
+            if (serializedElement !== undefined) {
+              outputArray.push(serializedElement);
+            }
+          }
+          return outputArray;
+        }
+
+        const outputObject: object = { };
+        for (const key of Object.getOwnPropertyNames(input)) {
+          // tslint:disable-next-line:no-any
+          const value: any = input[key];
+
+          // tslint:disable-next-line:no-any
+          const serializedValue: any = MessageRouter.buildJsonDumpObject(value);
+
+          if (serializedValue !== undefined) {
+            outputObject[key] = serializedValue;
+          }
+        }
+        return outputObject;
+    }
+
+    return undefined;
```

Comment:
Do you want to return something more useful if the parameter is a function. Maybe `<function>`?

Reply:
Not really.  The idea is that this function crawls through an object and picks out the printable stuff, ignoring any other members.

Ideally I would have just used `JSON.serialize()`; however, we need to dump the object returned by TypeScript's `ts.program.getCompilerOptions()`.  I was paranoid that future compiler releases might add some methods to that object.